packages/job-worker/src/ingest/bucketAdlibs.ts

```diff
@@ -134,155 +153,161 @@ export async function handleBucketItemImport(context: JobContext, data: BucketIt
-
-		const watchedPackages = WatchedPackagesHelper.empty(context)
-
-		const contextForVariant = new ShowStyleUserContext(
-			{
-				name: `Bucket Ad-Lib`,
-				identifier: `studioId=${context.studioId},showStyleBaseId=${showStyleBase._id},showStyleVariantId=${showStyleVariant._id}`,
-				tempSendUserNotesIntoBlackHole: true, // TODO-CONTEXT
-			},
-			context,
-			showStyleCompound,
-			watchedPackages
-		)
+	await context.directCollections.runInTransaction(async (transaction) => {
+		const ps: Promise<any>[] = []
+		let isFirstShowStyleVariant = true
+		for (const showStyleVariant of showStyleVariants) {
+			const showStyleCompound = createShowStyleCompound(showStyleBase, showStyleVariant)
+
+			if (!showStyleCompound)
+				throw new Error(
+					`Unable to create a ShowStyleCompound for ${showStyleBase._id}, ${showStyleVariant._id} `
+				)
 
-		let rawAdlib: IBlueprintAdLibPiece | IBlueprintActionManifest | null = null
-		try {
-			if (blueprint.blueprint.getAdlibItem) {
-				rawAdlib = blueprint.blueprint.getAdlibItem(contextForVariant, data.payload)
+			const watchedPackages = WatchedPackagesHelper.empty(context)
+
+			const contextForVariant = new ShowStyleUserContext(
+				{
+					name: `Bucket Ad-Lib`,
+					identifier: `studioId=${context.studioId},showStyleBaseId=${showStyleBase._id},showStyleVariantId=${showStyleVariant._id}`,
+					tempSendUserNotesIntoBlackHole: true, // TODO-CONTEXT
+				},
+				context,
+				showStyleCompound,
+				watchedPackages
+			)
+
+			let rawAdlib: IBlueprintAdLibPiece | IBlueprintActionManifest | null = null
+			try {
+				if (blueprint.blueprint.getAdlibItem) {
+					rawAdlib = blueprint.blueprint.getAdlibItem(contextForVariant, data.payload)
+				}
+			} catch (err) {
+				logger.error(`Error in showStyleBlueprint.getShowStyleVariantId: ${stringifyError(err)}`)
+				rawAdlib = null
 			}
-		} catch (err) {
-			logger.error(`Error in showStyleBlueprint.getShowStyleVariantId: ${stringifyError(err)}`)
-			rawAdlib = null
-		}
 
-		const importVersions: RundownImportVersions = {
-			studio: studio._rundownVersionHash,
-			showStyleBase: showStyleCompound._rundownVersionHash,
-			showStyleVariant: showStyleCompound._rundownVersionHashVariant,
-			blueprint: blueprint.blueprint.blueprintVersion,
-			core: getSystemVersion(),
-		}
+			const importVersions: RundownImportVersions = {
+				studio: studio._rundownVersionHash,
+				showStyleBase: showStyleCompound._rundownVersionHash,
+				showStyleVariant: showStyleCompound._rundownVersionHashVariant,
+				blueprint: blueprint.blueprint.blueprintVersion,
+				core: getSystemVersion(),
+			}
 
-		if (rawAdlib) {
-			// Cache the newRank, so we only have to calculate it once:
-			if (newRank === undefined) {
-				const [highestAdlib, highestAction] = await Promise.all([
-					context.directCollections.BucketAdLibPieces.findFetch(
-						{
-							bucketId: data.bucketId,
-						},
-						{
-							sort: {
-								_rank: -1,
+			if (rawAdlib) {
```

Comment:
If this loop body were extracted to a separate function, it could also help with this huge condition block, and just have it abort early.